Enable In-Memory Analytics Accelerator to accelerate data compression/decompression

Jiang-Hua opened the pull request:
### What problem does this PR solve?

Issue Number: close #7329

Problem Summary:

### What is changed and how it works?
In-Memory Analytics Accelerator(IAA) is a hardware accelerator embedded in CPU, it delivered in 4th gen Intel Xeon Scalable processors (code name: SPR, Launched in January 2023). IAA integrated accelerator IP that accelerates analytics primitives (scan, filter, etc.), CRC calculations, compression, decompression, and more.

Query Processing Library (QPL) is an open-source library which can offload data compression/decompression to IAA hardware.

LZ4_decompress_safe is one of top hotsport for TiFlash. IAA/QPL compression can replace the current compression algorithm (LZ4 or ZSTD) and improve performance and compression ratio, reduce CPU usage.

### Check List

Tests <!-- At least one of them must be included. -->

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
QPL has 2 paths to compress/decompress data. One is hardware path which offloads compress/decompress to IAA hardware. Another one is software path, when the system doesn't have IAA hardware, QPL will use CPU and AVX512 to compress/decompress data. The data compressed by hardware/software path are compatible.
1. Software path test:
Modify tiflash-deploy/conf/ tiflash.toml and add the following parameter:
[profiles.default]
dt_compression_method = "qpl"
Restart tiflash and run tiup bench tpch
2. Hardware path test:
Because IAA hardware just released in January 2023. So most kernels do not support IAA yet. We need 5.18+ kernel and change some kernel parameters. This operation may not be necessary in the future.
CONFIG_INTEL_IDXD_BUS=y
CONFIG_INTEL_IDXD=m
CONFIG_INTEL_IDXD_COMPAT=y
CONFIG_INTEL_IDXD_SVM=y
CONFIG_INTEL_IDXD_PERFMON=y
After rebuilding kernel, we need to download a IAA config tool https://github.com/intel/idxd-config.git . It is used to enable/disable IAA accelerators.
The following steps are same to the software path test.
Modify tiflash-deploy/conf/ tiflash.toml and add the following parms
[profiles.default]
dt_compression_method = "qpl"
Restart tiflash and run tiup bench tpch
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [x] Contains variable changes
tiflash.toml
[profiles.default]
dt_compression_method = "qpl"
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

<!-- bugfix or new feature needs a release note -->

```release-note
Support In-Memory Analytics Accelerator(IAA) and Query Processing Library (QPL) to accelerate data compression/decompression
```

Jiang-Hua commented:
The build error is
[2023-04-24T08:16:16.951Z] -- Intel QPL version: 1.1.0
[2023-04-24T08:16:16.951Z] -- Looking for a ASM_NASM compiler
[2023-04-24T08:16:17.208Z] -- Looking for a ASM_NASM compiler - NOTFOUND
[2023-04-24T08:16:17.208Z] CMake Error at contrib/qpl-cmake/CMakeLists.txt:39 (message):
[2023-04-24T08:16:17.208Z]   Please install NASM from 'https://www.nasm.us/' because NASM compiler can
[2023-04-24T08:16:17.208Z]   not be found!

ASM_NASM compiler is required software packages for QPL. We usually use the package management tool such as yum/dnf/apt/deb to install it.

How should I fix the build error?
Put ASM_NASM into "Prepare Prerequisites"?
Download ASM_NASM package and compile it automatically in the build?

Lloyd-Pottiger commented:
> The build error is [2023-04-24T08:16:16.951Z] -- Intel QPL version: 1.1.0 [2023-04-24T08:16:16.951Z] -- Looking for a ASM_NASM compiler [2023-04-24T08:16:17.208Z] -- Looking for a ASM_NASM compiler - NOTFOUND [2023-04-24T08:16:17.208Z] CMake Error at contrib/qpl-cmake/CMakeLists.txt:39 (message): [2023-04-24T08:16:17.208Z] Please install NASM from '[nasm.us](https://www.nasm.us/)' because NASM compiler can [2023-04-24T08:16:17.208Z] not be found!
> 
> ASM_NASM compiler is required software packages for QPL. We usually use the package management tool such as yum/dnf/apt/deb to install it.
> 
> How should I fix the build error? Put ASM_NASM into "Prepare Prerequisites"? Download ASM_NASM package and compile it automatically in the build?

Seems that our ci env is too old (centos 7) 😢 , maybe we can add a variable to control whether to compile ppl?

zanmato1984 commented:
Thanks for the contribution @Jiang-Hua . It's much appreciated.

Before going through details in this PR, I want to align our goals with you first:

1. A cmake variable `ENABLE_QPL` will be used to control build with qpl or not, default to true for x86 platforms with avx512 support, false otherwise.
2. In runtime, when qpl is specified by user, i.e., in tiflash config file, an explicit error should be reported if the binary is not built with qpl.
3. Even if the binary is built with qpl but the actual hardware doesn't have iaa (e.g., via some runtime check), I recommend to give an explicit error as well, or at least a warning in the log, as the performance is expected to be suboptimal.
4. Our CI build will cover x86 (with avx512) by default, so the sanity of this PR will be checked in CI.
5. Though I don't think our test machines have CPUs with iaa support, some corresponding tests are still required. We can talk about the hardware for testing in the future.

If we have agreement regarding to above, I'll further move forward into some specifics about the build failure. The problem is that we don't have asm_nasm package in our builder image (a docker image that contains both the standard OS and required 3rd party libraries). And according to @Lloyd-Pottiger , our base OS is centos 7 and it's too old to install the latest asm_nasm. Plus we have other OS-es such as Rocky Linux, etc. Maybe our best shot is that you supply us the pre-built asm-nasm libraries on all OS-es we support, probably in the form of docker images, from which we can easily copy the library into our builder images.

Looking forward to your feedback. Thanks.

zanmato1984 commented on CMakeLists.txt:
```diff
@@ -302,6 +302,11 @@ include (cmake/find_prometheus.cmake)
 include (cmake/find_tiflash_proxy.cmake)
 include (cmake/find_xxhash.cmake)
 
+if (USE_QPL AND OS_LINUX AND ARCH_AMD64 AND TIFLASH_ENABLE_AVX512_SUPPORT)
```
Actually we don't actually need `USE_QPL`. `ENABLE_QPL` should be sufficient.

Suggest changes followed.

zanmato1984 commented on CMakeLists.txt:
```diff
@@ -302,6 +302,11 @@ include (cmake/find_prometheus.cmake)
 include (cmake/find_tiflash_proxy.cmake)
 include (cmake/find_xxhash.cmake)
 
+if (USE_QPL AND OS_LINUX AND ARCH_AMD64 AND TIFLASH_ENABLE_AVX512_SUPPORT)
+    option (ENABLE_QPL "Enable Intel Query Processing Library" ON)
+    include (cmake/find_qpl.cmake)
+endif()
```
```suggestion
if (OS_LINUX AND ARCH_AMD64 AND TIFLASH_ENABLE_AVX512_SUPPORT)
    set (ENABLE_QPL_DEFAULT ON)
else ()
    set (ENABLE_QPL_DEFAULT ON)
endif()
option (ENABLE_QPL "Enable Intel Query Processing Library" ${ENABLE_QPL_DEFAULT})

if (ENABLE_QPL)
    if (OS_LINUX AND ARCH_AMD64 AND TIFLASH_ENABLE_AVX512_SUPPORT)
        include (cmake/find_qpl.cmake)
    else ()
        message (FATAL_ERROR "`ENABLE_QPL` is not supported on this platform")
    endif ()
endif ()
```

Jiang-Hua replied:
Line 308 should be "set (ENABLE_QPL_DEFAULT OFF)", right?
The other parts are good.

zanmato1984 replied:
It is, I made a modification minutes after my original change. Maybe your page is outdated.

But by looking again I just realized two things:
1. This logic could be further simplified.
2. For developers who need to build tiflash, we don't want to enforce using qpl for them, i.e., require asm_nasm installed in the system. So maybe it's more appropriate to default this option to false. Consequently, we'll explicitly enable this option in our official build, which will be done in other scripts. I'll update the script later.

Jiang-Hua replied:
Looks good to me.

zanmato1984 commented on contrib/CMakeLists.txt:
```diff
@@ -16,6 +16,11 @@ if (USE_INTERNAL_ZSTD_LIBRARY)
+if (USE_INTERNAL_QPL_LIBRARY)
+    add_subdirectory (qpl-cmake)
+endif ()
+
```
```suggestion
```

Jiang-Hua replied:
Commited, thanks.

zanmato1984 commented:
> Thanks for the contribution @Jiang-Hua . It's much appreciated.
> 
> Before going through details in this PR, I want to align our goals with you first:
> 
> 1. A cmake variable `ENABLE_QPL` will be used to control build with qpl or not, default to true for x86 platforms with avx512 support, false otherwise.
> 2. In runtime, when qpl is specified by user, i.e., in tiflash config file, an explicit error should be reported if the binary is not built with qpl.
> 3. Even if the binary is built with qpl but the actual hardware doesn't have iaa (e.g., via some runtime check), I recommend to give an explicit error as well, or at least a warning in the log, as the performance is expected to be suboptimal.
> 4. Our CI build will cover x86 (with avx512) by default, so the sanity of this PR will be checked in CI.
> 5. Though I don't think our test machines have CPUs with iaa support, some corresponding tests are still required. We can talk about the hardware for testing in the future.
> 
> If we have agreement regarding to above, I'll further move forward into some specifics about the build failure. The problem is that we don't have asm_nasm package in our builder image (a docker image that contains both the standard OS and required 3rd party libraries). And according to @Lloyd-Pottiger , our base OS is centos 7 and it's too old to install the latest asm_nasm. Plus we have other OS-es such as Rocky Linux, etc. Maybe our best shot is that you supply us the pre-built asm-nasm libraries on all OS-es we support, probably in the form of docker images, from which we can easily copy the library into our builder images.
> 
> Looking forward to your feedback. Thanks.

Alternatively, you can as well upload asm_nasm for older OS-es, such as centos 7, into some package sources that we can access.

Jiang-Hua commented:
Thanks for the response.

> Thanks for the contribution @Jiang-Hua . It's much appreciated.
> 
> Before going through details in this PR, I want to align our goals with you first:
> 
> 1. A cmake variable `ENABLE_QPL` will be used to control build with qpl or not, default to true for x86 platforms with avx512 support, false otherwise.

Got it.

> 2. In runtime, when qpl is specified by user, i.e., in tiflash config file, an explicit error should be reported if the binary is not built with qpl.

We used "#ifdef ENABLE_QPL_COMPRESSION" to control whether the relevant code is compiled. A exception will be thrown for the above case in SettingCompressionMethod::getCompressionMethod, is that ok？

dbms/src/Interpreters/SettingsCommon.h :
        if (lower_str == "zstd")
            return CompressionMethod::ZSTD;
#ifdef ENABLE_QPL_COMPRESSION
        if (lower_str == "qpl")
            return CompressionMethod::QPL;
#endif

        throw Exception("Unknown compression method: '" + s + "', must be one of 'lz4', 'lz4hc', 'zstd', 'qpl'", ErrorCodes::UNKNOWN_COMPRESSION_METHOD);

There is a mistake for the error message, we should remove qpl when the binary is not built with qpl. Will fix it.

> 3. Even if the binary is built with qpl but the actual hardware doesn't have iaa (e.g., via some runtime check), I recommend to give an explicit error as well, or at least a warning in the log, as the performance is expected to be suboptimal.

Currently, if the actual hardware doesn't have iaa, QPL will use CPU and AVX512 to compress/decompress data and without any message output.
Good suggestion. Will add a warning.

> 4. Our CI build will cover x86 (with avx512) by default, so the sanity of this PR will be checked in CI.

Ok.

> 5. Though I don't think our test machines have CPUs with iaa support, some corresponding tests are still required. We can talk about the hardware for testing in the future.

Yes, we can provide a testing environment with IAA.

> 
> If we have agreement regarding to above, I'll further move forward into some specifics about the build failure. The problem is that we don't have asm_nasm package in our builder image (a docker image that contains both the standard OS and required 3rd party libraries). And according to @Lloyd-Pottiger , our base OS is centos 7 and it's too old to install the latest asm_nasm. Plus we have other OS-es such as Rocky Linux, etc. Maybe our best shot is that you supply us the pre-built asm-nasm libraries on all OS-es we support, probably in the form of docker images, from which we can easily copy the library into our builder images.

The website of nasm provides binary libraries - https://www.nasm.us/pub/nasm/releasebuilds/2.16.01/ .
I think most OS-es can install nasm via package management tools such as yum/dnf/apt/deb, this method is safer and more convenient.
Anyway, if the above methods are not feasible, please provide the OS and version you require, we will try to find binary libraries for them.

Thanks,
Hua
> 
> Looking forward to your feedback. Thanks.

zanmato1984 commented:
> Thanks for the response.
> 
> > Thanks for the contribution @Jiang-Hua . It's much appreciated.
> > Before going through details in this PR, I want to align our goals with you first:
> > 
> > 1. A cmake variable `ENABLE_QPL` will be used to control build with qpl or not, default to true for x86 platforms with avx512 support, false otherwise.
> 
> Got it.

As described in my other comment, will default to false.

> 
> > 2. In runtime, when qpl is specified by user, i.e., in tiflash config file, an explicit error should be reported if the binary is not built with qpl.
> 
> We used "#ifdef ENABLE_QPL_COMPRESSION" to control whether the relevant code is compiled. A exception will be thrown for the above case in SettingCompressionMethod::getCompressionMethod, is that ok？
> 

That sounds good.

> dbms/src/Interpreters/SettingsCommon.h : if (lower_str == "zstd") return CompressionMethod::ZSTD; #ifdef ENABLE_QPL_COMPRESSION if (lower_str == "qpl") return CompressionMethod::QPL; #endif
> 
> ```
>     throw Exception("Unknown compression method: '" + s + "', must be one of 'lz4', 'lz4hc', 'zstd', 'qpl'", ErrorCodes::UNKNOWN_COMPRESSION_METHOD);
> ```
> 
> There is a mistake for the error message, we should remove qpl when the binary is not built with qpl. Will fix it.
> 

You are right. But I haven't reviewed that far. Go ahead with the fix and I'll look at the code later.

> > 3. Even if the binary is built with qpl but the actual hardware doesn't have iaa (e.g., via some runtime check), I recommend to give an explicit error as well, or at least a warning in the log, as the performance is expected to be suboptimal.
> 
> Currently, if the actual hardware doesn't have iaa, QPL will use CPU and AVX512 to compress/decompress data and without any message output. Good suggestion. Will add a warning.
> 

OK.

> > 4. Our CI build will cover x86 (with avx512) by default, so the sanity of this PR will be checked in CI.
> 
> Ok.
> 
> > 5. Though I don't think our test machines have CPUs with iaa support, some corresponding tests are still required. We can talk about the hardware for testing in the future.
> 
> Yes, we can provide a testing environment with IAA.
> 

Will open a following thread about this. Let this long paragraph discussion end here.

> > If we have agreement regarding to above, I'll further move forward into some specifics about the build failure. The problem is that we don't have asm_nasm package in our builder image (a docker image that contains both the standard OS and required 3rd party libraries). And according to @Lloyd-Pottiger , our base OS is centos 7 and it's too old to install the latest asm_nasm. Plus we have other OS-es such as Rocky Linux, etc. Maybe our best shot is that you supply us the pre-built asm-nasm libraries on all OS-es we support, probably in the form of docker images, from which we can easily copy the library into our builder images.
> 
> The website of nasm provides binary libraries - [nasm.us/pub/nasm/releasebuilds/2.16.01](https://www.nasm.us/pub/nasm/releasebuilds/2.16.01/) . I think most OS-es can install nasm via package management tools such as yum/dnf/apt/deb, this method is safer and more convenient. Anyway, if the above methods are not feasible, please provide the OS and version you require, we will try to find binary libraries for them.
> 

Yeah, we'll do some experiment regarding to the package by ourselves and come to you if we have any problem.

> Thanks, Hua
> 
> > Looking forward to your feedback. Thanks.

Thanks again.

zanmato1984 commented:
> > 5. Though I don't think our test machines have CPUs with iaa support, some corresponding tests are still required. We can talk about the hardware for testing in the future.
> 
> Yes, we can provide a testing environment with IAA.
> 

Sorry I didn't make myself clear in my original comment. Let me elaborate:
1. The test code is required no matter if we have the IAA machine or not. That means you'll have to add some tests in our code base: UT (gtests), or integration tests (https://github.com/pingcap/tiflash/tree/master/tests), or both - and these tests will be running in CI as you add them into your PR.
2. We don't have IAA machines so far. It'll be nice if we can leverage your environment. But one thing to note is that we are expecting some environment long-termly available, because we want these tests to run regularly in our CI (for PR), daily regression, and production release testing. We may need the environment like forever, until we have our own IAA machines someday. Does it sound reasonable to you? 

Thanks.

Jiang-Hua commented:
> 1. The test code is required no matter if we have the IAA machine or not. That means you'll have to add some tests in our code base: UT (gtests), or integration tests (https://github.com/pingcap/tiflash/tree/master/tests), or both - and these tests will be running in CI as you add them into your PR.

Ok, we will add some tests for this feature.

> 2. We don't have IAA machines so far. It'll be nice if we can leverage your environment. But one thing to note is that we are expecting some environment long-termly available, because we want these tests to run regularly in our CI (for PR), daily regression, and production release testing. We may need the environment like forever, until we have our own IAA machines someday. Does it sound reasonable to you?
> 
I'm not sure if we can provide a long-term IAA environment, I need to discuss this with my colleagues in the marketing department. The good news is that most of the new Xeon servers this year should have IAA accelerator.

zanmato1984 commented on dbms/src/IO/CodecDeflateQpl.h:
```diff
@@ -0,0 +1,82 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef SRC_LIB_COMPRESS_QPL_H_
```
We prefer `#pragma once` instead of this manual protection.

zanmato1984 commented on dbms/src/IO/CodecDeflateQpl.h:
```diff
@@ -0,0 +1,82 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef SRC_LIB_COMPRESS_QPL_H_
```
I think this header needs to be enclosed by `#ifdef USE_QPL_COMPRESSION`, as it is directly included in [CompressedReadBufferBase.cpp](https://github.com/pingcap/tiflash/pull/7364/files#diff-7d863f62031fa22aa0c5a5cad85ced3bf9b5344fb09ef96c27c023a0cf4eab21).

Jiang-Hua replied:
Done，thanks.

zanmato1984 commented:
I think current way of organizing qpl-related code, including both compile time and runtime, is not quite unified. For example `CompressionMethodByte::QPL` is a valid symbol even not building with qpl, but some qpl branch is totally invisible (thus the error handling is by falling through to the final "unknown compression method" part).

We should either allow "qpl" enum and string const (in error message) w/ or w/o qpl built, and for any reference of these enum and string const, report explicit error for non-qpl build:
```
#ifndef USE_QPL_COMPRESSION
  throw exception("This binary is not built with qpl support");
#endif
```
Or, totally hide any "qpl" thing, including enum, string const, actual branches dealing with qpl compression/decompression, for a non-qpl build.

Personally I prefer the former a bit.

zanmato1984 commented on dbms/src/Interpreters/SettingsCommon.h:
```diff
@@ -772,15 +772,19 @@ struct SettingCompressionMethod
             return CompressionMethod::LZ4HC;
         if (lower_str == "zstd")
             return CompressionMethod::ZSTD;
+#ifdef ENABLE_QPL_COMPRESSION
```
This is one weird thing of what I mentioned in my other comment. If user specifies "qpl" in config file but uses a non-qpl binary, the user sees a error message `Unknown compression method: 'qpl', must be one of 'lz4', 'lz4hc', 'zstd', 'qpl'` - pretty confusing.

zanmato1984 replied:
Hope this helps explaining what I meant by we either ... or ...

Jiang-Hua commented:
> We should either allow "qpl" enum and string const (in error message) for qpl and non-qpl build, and when referencing any of these enum and string const, report explicit error for non-qpl build:
> 
> ```
> #ifndef USE_QPL_COMPRESSION
>   throw exception("This binary is not built with qpl support");
> #endif
> ```

I agree with you. We have updated codes based on your above suggestion.

zanmato1984 commented:
I see that `ENABLE_QPL_COMPRESSION` is being used as the macro for cpp code. But it is only defined in https://github.com/pingcap/tiflash/pull/7364/files#diff-3e9d229a7049e3d3b0ef0ba9b728e6dfa6620bb254973d467ede8d439f37032aR314, which is only within "internal" qpl library. Considering we also support enabling QPL via "external" qpl library (i.e., pre-installed by the system), this macro won't be available in such case, resulting in QPL code in cpp not compiled.

I suggest that:
1. Change the macro name to `USE_QPL`, which complies the naming convention of other compiler switch macros.
2. Set a cmake variable `USE_QPL` in `cmake/find_qpl.cmake`, as long as the libraries are successfully found, regardless of it's internal or external. 
3. Turn the cmake variable `USE_QPL` into macro `USE_QPL` via these two files (https://github.com/pingcap/tiflash/blob/4ecf3cd7649f0597c540cdbfe4ea609f26d7d169/dbms/src/Common/config.h.in, https://github.com/pingcap/tiflash/blob/4ecf3cd7649f0597c540cdbfe4ea609f26d7d169/dbms/src/Common/config_build.cpp.in).

Tip: You can search for `USE_ICU` as an example.

@Jiang-Hua

zanmato1984 commented on dbms/src/Common/ErrorCodes.cpp:
```diff
@@ -394,6 +394,11 @@ extern const int UNSUPPORTED_URI_SCHEME = 450;
 extern const int UNACCEPTABLE_URL = 450;
 extern const int TOO_MANY_REDIRECTS = 450;
 
+extern const int QPL_INIT_JOB_FAILED = 453;
+extern const int QPL_ACQUIRE_JOB_FAILED = 454;
+extern const int QPL_COMPRESS_DATA_FAILED = 455;
+extern const int QPL_DECOMPRESS_DATA_FAILED = 456;
```
We may want to enclose these defines by the `USE_QPL` macro.

zanmato1984 commented on dbms/src/IO/CompressedStream.h:
```diff
@@ -31,7 +31,8 @@ enum class CompressionMethod
     LZ4HC = 2, /// The format is the same as for LZ4. The difference is only in compression.
     ZSTD = 3, /// Experimental algorithm: https://github.com/Cyan4973/zstd
-    NONE = 4, /// No compression
+    QPL = 4, /// The Intel Query Processing Library (QPL) is an open-source library to provide high-performance query processing operations
```
We may want to enclose this enum by the `USE_QPL` macro.

zanmato1984 commented on dbms/src/IO/CompressedStream.h:
```diff
@@ -60,6 +61,7 @@ enum class CompressionMethodByte : uint8_t
     NONE = 0x02,
     LZ4 = 0x82,
     ZSTD = 0x90,
+    QPL = 0x88,
```
Ditto.

zanmato1984 commented on dbms/src/Interpreters/SettingsCommon.h:
```diff
@@ -772,15 +772,21 @@ struct SettingCompressionMethod
             return CompressionMethod::LZ4HC;
         if (lower_str == "zstd")
             return CompressionMethod::ZSTD;
+        if (lower_str == "qpl")
+#ifndef ENABLE_QPL_COMPRESSION
+            throw exception("This binary is not built with qpl support", ErrorCodes::UNKNOWN_COMPRESSION_METHOD);
+#else
+            return CompressionMethod::QPL;
+#endif
```
```suggestion
#ifdef ENABLE_QPL_COMPRESSION
        if (lower_str == "qpl")
            return CompressionMethod::QPL;
#endif
```

zanmato1984 commented on dbms/src/Interpreters/SettingsCommon.h:
```diff
@@ -772,15 +772,21 @@ struct SettingCompressionMethod
+#endif
 
-        throw Exception("Unknown compression method: '" + s + "', must be one of 'lz4', 'lz4hc', 'zstd'", ErrorCodes::UNKNOWN_COMPRESSION_METHOD);
+        throw Exception("Unknown compression method: '" + s + "', must be one of 'lz4', 'lz4hc', 'zstd', 'qpl'", ErrorCodes::UNKNOWN_COMPRESSION_METHOD);
```
```suggestion
#ifdef USE_QPL
        throw Exception("Unknown compression method: '" + s + "', must be one of 'lz4', 'lz4hc', 'zstd', 'qpl'", ErrorCodes::UNKNOWN_COMPRESSION_METHOD);
#else
        throw Exception("Unknown compression method: '" + s + "', must be one of 'lz4', 'lz4hc', 'zstd'", ErrorCodes::UNKNOWN_COMPRESSION_METHOD);
#endif
```

zanmato1984 commented on dbms/src/Interpreters/SettingsCommon.h:
```diff
@@ -772,15 +772,21 @@ struct SettingCompressionMethod
     String toString() const
     {
-        const char * strings[] = {nullptr, "lz4", "lz4hc", "zstd"};
+        const char * strings[] = {nullptr, "lz4", "lz4hc", "zstd", "qpl"};
```
```suggestion
#ifdef USE_QPL
        const char * strings[] = {nullptr, "lz4", "lz4hc", "zstd", "qpl"};
        auto compression_method_last = CompressionMethod::QPL;
#else
        const char * strings[] = {nullptr, "lz4", "lz4hc", "zstd"};
        auto compression_method_last = CompressionMethod::ZSTD;
#endif
```

zanmato1984 commented on dbms/src/Interpreters/SettingsCommon.h:
```diff
@@ -772,15 +772,21 @@ struct SettingCompressionMethod
+        const char * strings[] = {nullptr, "lz4", "lz4hc", "zstd", "qpl"};
 
-        if (value < CompressionMethod::LZ4 || value > CompressionMethod::ZSTD)
+        if (value < CompressionMethod::LZ4 || value > CompressionMethod::QPL)
```
```suggestion
        if (value < CompressionMethod::LZ4 || value > compression_method_last)
```

zanmato1984 commented on dbms/src/Server/DTTool/DTToolMigrate.cpp:
```diff
@@ -58,7 +58,7 @@ static constexpr char MIGRATE_HELP[] =
     "  --algorithm   Checksum algorithm. [default: xxh3] [available: xxh3, city128, crc32, crc64, none]\n"
     "  --frame       Checksum frame length. [default: " TO_STRING(TIFLASH_DEFAULT_CHECKSUM_FRAME_SIZE) "]\n"
-    "  --compression Compression method. [default: lz4] [available: lz4, lz4hc, zstd, none]\n"
+    "  --compression Compression method. [default: lz4] [available: lz4, lz4hc, zstd, qpl, none]\n"
```
```suggestion
#ifdef USE_QPL
    "  --compression Compression method. [default: lz4] [available: lz4, lz4hc, zstd, qpl, none]\n"
#else
    "  --compression Compression method. [default: lz4] [available: lz4, lz4hc, zstd, none]\n"
#endif
```

Lloyd-Pottiger commented on cmake/find_qpl.cmake:
```diff
@@ -0,0 +1,34 @@
+# Copyright 2022 PingCAP, Ltd.
```
```suggestion
# Copyright 2023 PingCAP, Ltd.
```

Lloyd-Pottiger commented on dbms/CMakeLists.txt:
```diff
@@ -253,6 +254,11 @@ endif ()
 if (NOT USE_INTERNAL_ZSTD_LIBRARY)
     target_include_directories (dbms BEFORE PRIVATE ${ZSTD_INCLUDE_DIR})
 endif ()
+if (ENABLE_QPL)
+    if (NOT USE_INTERNAL_QPL_LIBRARY)
+        target_include_directories (dbms BEFORE PRIVATE ${QPL_INCLUDE_DIR})
+    endif ()
+endif ()
```
```suggestion
if (ENABLE_QPL AND NOT USE_INTERNAL_QPL_LIBRARY)
    target_include_directories (dbms BEFORE PRIVATE ${QPL_INCLUDE_DIR})
endif ()
```

Lloyd-Pottiger commented on dbms/src/IO/CodecDeflateQpl.h:
```diff
@@ -0,0 +1,82 @@
+// Copyright 2022 PingCAP, Ltd.
```
```suggestion
// Copyright 2023 PingCAP, Ltd.
```

Lloyd-Pottiger commented on dbms/src/IO/CodecDeflateQpl.cpp:
```diff
@@ -0,0 +1,238 @@
+#ifdef ENABLE_QPL_COMPRESSION
```
add license header

Lloyd-Pottiger commented on dbms/src/IO/CodecDeflateQpl.h:
```diff
@@ -0,0 +1,82 @@
+// limitations under the License.
+
+#pragma once
+#ifdef ENABLE_QPL_COMPRESSION
```
sorry, where have defined the macro `ENABLE_QPL_COMPRESSION` ?

zanmato1984 replied:
See my comment: https://github.com/pingcap/tiflash/pull/7364#issuecomment-1605860762

Lloyd-Pottiger commented on dbms/src/IO/CodecDeflateQpl.h:
```diff
@@ -0,0 +1,82 @@
+#include <memory>
+#include <random>
+
+#include "qpl/qpl.h"
```
```suggestion
#include <qpl/qpl.h>
```

hongyunyan commented on dbms/src/IO/CodecDeflateQpl.cpp:
```diff
@@ -0,0 +1,238 @@
+                delete[] jobs_buffer;
+                if (path == qpl_path_hardware)
+                {
+                    throw Exception(std::string("Initialization of IAA hardware failed:") + std::to_string(status) + " will attempt to use software DeflateQpl codec instead of hardware DeflateQpl codec.", ErrorCodes::QPL_INIT_JOB_FAILED);
```
fmt::format could be a better choice here

hongyunyan commented on dbms/src/IO/CodecDeflateQpl.cpp:
```diff
@@ -0,0 +1,238 @@
+                }
+                else
+                {
+                    throw Exception(std::string("Initialization of software DeflateQpl codec failed:") + std::to_string(status) + " QPL compression/decompression cannot be enabled.", ErrorCodes::QPL_INIT_JOB_FAILED);
```
Ditto

Jiang-Hua commented:
> I see that `ENABLE_QPL_COMPRESSION` is being used as the macro for cpp code. But it is only defined in https://github.com/pingcap/tiflash/pull/7364/files#diff-3e9d229a7049e3d3b0ef0ba9b728e6dfa6620bb254973d467ede8d439f37032aR314, which is only within "internal" qpl library. Considering we also support enabling QPL via "external" qpl library (i.e., pre-installed by the system), this macro won't be available in such case, resulting in QPL code in cpp not compiled.
> 
> I suggest that:
> 
> 1. Change the macro name to `USE_QPL`, which complies the naming convention of other compiler switch macros.

Thanks for reviewing.
There are 2 variables ENABLE_QPL (https://github.com/pingcap/tiflash/pull/7364#issuecomment-1521452861) and USE_QPL. Only replace ENABLE_QPL_COMPRESSION with USE_QPL.   ENABLE_QPL remains unchanged, right?

> 2. Set a cmake variable `USE_QPL` in `cmake/find_qpl.cmake`, as long as the libraries are successfully found, regardless of it's internal or external.

The original logical was copied from find_lz4.cmake, which is a little bit confusing. We rewrote find_qpl.cmake, if the internal QPL or system QPL is not found, a error message will be shown and compilation will be terminated.
If find_qpl.cmake passes, USE_QPL will be set to ON.

> 3. Turn the cmake variable `USE_QPL` into macro `USE_QPL` via these two files (https://github.com/pingcap/tiflash/blob/4ecf3cd7649f0597c540cdbfe4ea609f26d7d169/dbms/src/Common/config.h.in, https://github.com/pingcap/tiflash/blob/4ecf3cd7649f0597c540cdbfe4ea609f26d7d169/dbms/src/Common/config_build.cpp.in).
> 
> Tip: You can search for `USE_ICU` as an example.

I don't quite know how they work, but I have added the definition of USE_QPL in config.h.in and config_build.cpp.in like USE_ICU . Please help to review them.

> 
> @Jiang-Hua

Jiang-Hua commented:
I updated the code according to above suggestions, please help to review them, thanks a lot.

zanmato1984 commented:
> > I see that `ENABLE_QPL_COMPRESSION` is being used as the macro for cpp code. But it is only defined in [#7364 (files)](https://github.com/pingcap/tiflash/pull/7364/files#diff-3e9d229a7049e3d3b0ef0ba9b728e6dfa6620bb254973d467ede8d439f37032aR314), which is only within "internal" qpl library. Considering we also support enabling QPL via "external" qpl library (i.e., pre-installed by the system), this macro won't be available in such case, resulting in QPL code in cpp not compiled.
> > I suggest that:
> > 
> > 1. Change the macro name to `USE_QPL`, which complies the naming convention of other compiler switch macros.
> 
> Thanks for reviewing. There are 2 variables ENABLE_QPL ([#7364 (comment)](https://github.com/pingcap/tiflash/pull/7364#issuecomment-1521452861)) and USE_QPL. Only replace ENABLE_QPL_COMPRESSION with USE_QPL. ENABLE_QPL remains unchanged, right?
> 

Yes, you are right. A little bit context, our convention (though not strong), is that in cmake we use `ENABLE_XXX` to enable/disable a library for building. And if the library is enabled and successfully found, no matter internal or external (system), the cmake variable `USE_XXX` is set to `true`, which in turn becomes a definition passed to source code as a macro.

> > 2. Set a cmake variable `USE_QPL` in `cmake/find_qpl.cmake`, as long as the libraries are successfully found, regardless of it's internal or external.
> 
> The original logical was copied from find_lz4.cmake, which is a little bit confusing. We rewrote find_qpl.cmake, if the internal QPL or system QPL is not found, a error message will be shown and compilation will be terminated. If find_qpl.cmake passes, USE_QPL will be set to ON.
> 

I'll make some suggestion directly in the code.

> > 3. Turn the cmake variable `USE_QPL` into macro `USE_QPL` via these two files ([`4ecf3cd`/dbms/src/Common/config.h.in](https://github.com/pingcap/tiflash/blob/4ecf3cd7649f0597c540cdbfe4ea609f26d7d169/dbms/src/Common/config.h.in), [`4ecf3cd`/dbms/src/Common/config_build.cpp.in](https://github.com/pingcap/tiflash/blob/4ecf3cd7649f0597c540cdbfe4ea609f26d7d169/dbms/src/Common/config_build.cpp.in)).
> > 
> > Tip: You can search for `USE_ICU` as an example.
> 
> I don't quite know how they work, but I have added the definition of USE_QPL in config.h.in and config_build.cpp.in like USE_ICU . Please help to review them.
> 
> > @Jiang-Hua

Ditto.

zanmato1984 commented on CMakeLists.txt:
```diff
@@ -302,6 +302,19 @@ include (cmake/find_prometheus.cmake)
+if (ENABLE_QPL)
+    if (OS_LINUX AND ARCH_AMD64 AND TIFLASH_ENABLE_AVX512_SUPPORT)
+        include (cmake/find_qpl.cmake)
+	option(USE_QPL "on this platform supports the use of QPL" ON)
```
```suggestion
```

zanmato1984 replied:
`USE_XXX` is not supposed to be an option.

zanmato1984 commented on cmake/find_qpl.cmake:
```diff
@@ -0,0 +1,33 @@
+        message (FATAL_ERROR "System qpl library not found. Try to install system qpl library or set -DUSE_INTERNAL_QPL_LIBRARY=ON to use bundled qpl library.")
+    endif ()
+endif ()
+
```
```suggestion

if (QPL_LIBRARY AND QPL_INCLUDE_DIR)
    SET (USE_QPL 1)
endif ()
    
```

zanmato1984 commented on contrib/CMakeLists.txt:
```diff
@@ -16,6 +16,10 @@ if (USE_INTERNAL_ZSTD_LIBRARY)
     add_subdirectory (zstd-cmake)
 endif ()
 
+if (USE_INTERNAL_QPL_LIBRARY)
```
```suggestion
if (USE_QPL AND USE_INTERNAL_QPL_LIBRARY)
```

zanmato1984 commented on contrib/qpl-cmake/CMakeLists.txt:
```diff
@@ -0,0 +1,320 @@
+## The Intel® QPL provides high performance implementations of data processing functions for existing hardware accelerator, and/or software path in case if hardware accelerator is not available.
+if (NOT USE_QPL)
+    message(STATUS "Not using QPL")
+    return()
+ endif()
```
```suggestion
```

zanmato1984 commented on dbms/CMakeLists.txt:
```diff
@@ -123,6 +123,10 @@ add_library(tiflash_common_io ${SPLIT_SHARED} ${tiflash_common_io_headers} ${tif
 target_link_libraries(tiflash_common_io PUBLIC process_metrics)
 target_include_directories(tiflash_common_io BEFORE PUBLIC ${XXHASH_INCLUDE_DIR} ${TiFlash_SOURCE_DIR}/libs/libprocess_metrics/include)
 
+if (USE_QPL)
+    target_compile_definitions (tiflash_common_io PUBLIC USE_QPL)
+endif ()
+
```
```suggestion
```

Jiang-Hua replied:
We tried to remove these 3 lines, but USE_QPL cannot be passed to the compiler. Conditional compilation of USE_QPL will not work. So we restored them.

zanmato1984 replied:
I missed this comment. Withdrawing my lgtm. Let me take a look deeper.

zanmato1984 commented on dbms/CMakeLists.txt:
```diff
@@ -169,6 +173,7 @@ target_link_libraries (tiflash_common_io
     ${LINK_LIBRARIES_ONLY_ON_X86_64}
     ${LZ4_LIBRARY}
     ${ZSTD_LIBRARY}
+    ${QPL_LIBRARY}
```
```suggestion
```

zanmato1984 commented on dbms/CMakeLists.txt:
```diff
@@ -253,6 +258,9 @@ endif ()
 if (NOT USE_INTERNAL_ZSTD_LIBRARY)
     target_include_directories (dbms BEFORE PRIVATE ${ZSTD_INCLUDE_DIR})
 endif ()
+if (USE_QPL AND NOT USE_INTERNAL_QPL_LIBRARY)
+    target_include_directories (dbms BEFORE PRIVATE ${QPL_INCLUDE_DIR})
+endif ()
```
```suggestion

if (USE_QPL)
    target_link_libraries (dbms ${QPL_LIBRARY})
    target_include_directories (dbms PRIVATE ${QPL_INCLUDE_DIR})
endif ()
```

zanmato1984 commented on dbms/src/Common/ErrorCodes.cpp:
```diff
@@ -394,6 +394,13 @@ extern const int UNSUPPORTED_URI_SCHEME = 450;
 extern const int UNACCEPTABLE_URL = 450;
 extern const int TOO_MANY_REDIRECTS = 450;
 
+#ifdef USE_QPL
```
```suggestion
#if USE_QPL
```

zanmato1984 commented on dbms/src/IO/CodecDeflateQpl.cpp:
```diff
@@ -0,0 +1,252 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifdef USE_QPL
```
```suggestion
#if USE_QPL
```

zanmato1984 commented on dbms/src/IO/CompressedReadBufferBase.cpp:
```diff
@@ -119,6 +128,13 @@ void CompressedReadBufferBase<has_checksum>::decompress(char * to, size_t size_d
         if (ZSTD_isError(res))
             throw Exception("Cannot ZSTD_decompress: " + std::string(ZSTD_getErrorName(res)), ErrorCodes::CANNOT_DECOMPRESS);
     }
+#ifdef USE_QPL
```
```suggestion
#if USE_QPL
```

zanmato1984 commented on dbms/src/IO/CompressedStream.h:
```diff
@@ -31,7 +31,12 @@ enum class CompressionMethod
     LZ4 = 1,
     LZ4HC = 2, /// The format is the same as for LZ4. The difference is only in compression.
     ZSTD = 3, /// Experimental algorithm: https://github.com/Cyan4973/zstd
+#ifdef USE_QPL
```
```suggestion
#if USE_QPL
```

zanmato1984 commented on dbms/src/IO/CompressedStream.h:
```diff
@@ -60,6 +65,9 @@ enum class CompressionMethodByte : uint8_t
     NONE = 0x02,
     LZ4 = 0x82,
     ZSTD = 0x90,
+#ifdef USE_QPL
```
```suggestion
#if USE_QPL
```

zanmato1984 commented on dbms/src/IO/CompressedWriteBuffer.cpp:
```diff
@@ -23,6 +23,9 @@
 
 #include <memory>
 
+#ifdef USE_QPL
```
```suggestion
#if USE_QPL
```

zanmato1984 commented on dbms/src/IO/CompressedWriteBuffer.cpp:
```diff
@@ -111,6 +114,30 @@ size_t CompressionEncode(
 
         break;
     }
+#ifdef USE_QPL
```
```suggestion
#if USE_QPL
```

zanmato1984 commented on dbms/src/IO/CompressionSettings.cpp:
```diff
@@ -43,6 +43,10 @@ int CompressionSettings::getDefaultLevel(CompressionMethod method)
         return LZ4HC_CLEVEL_DEFAULT;
     case CompressionMethod::ZSTD:
         return 1;
+#ifdef USE_QPL
```
```suggestion
#if USE_QPL
```

zanmato1984 commented on dbms/src/Interpreters/SettingsCommon.h:
```diff
@@ -772,15 +772,26 @@ struct SettingCompressionMethod
         if (lower_str == "zstd")
             return CompressionMethod::ZSTD;
-
+#ifdef USE_QPL
```
```suggestion
#if USE_QPL
```

zanmato1984 commented on dbms/src/Interpreters/SettingsCommon.h:
```diff
@@ -772,15 +772,26 @@ struct SettingCompressionMethod
 
     String toString() const
     {
+#ifdef USE_QPL
```
```suggestion
#if USE_QPL
```

zanmato1984 commented on dbms/src/Server/DTTool/DTToolMigrate.cpp:
```diff
@@ -58,7 +58,11 @@ static constexpr char MIGRATE_HELP[] =
     "  --version     Target dtfile version. [default: 2] [available: 1, 2]\n"
     "  --algorithm   Checksum algorithm. [default: xxh3] [available: xxh3, city128, crc32, crc64, none]\n"
     "  --frame       Checksum frame length. [default: " TO_STRING(TIFLASH_DEFAULT_CHECKSUM_FRAME_SIZE) "]\n"
+#ifdef USE_QPL
```
```suggestion
#if USE_QPL
```

zanmato1984 commented on dbms/src/Server/DTTool/DTToolMigrate.cpp:
```diff
@@ -335,6 +339,12 @@ int migrateEntry(const std::vector<std::string> & opts, RaftStoreFFIFunc ffi_fun
             {
                 args.compression_method = DB::CompressionMethod::ZSTD;
             }
+#ifdef USE_QPL
```
```suggestion
#if USE_QPL
```

zanmato1984 commented:
/rebuild

zanmato1984 commented:
I checked the latest CI build. The cpp build seems fine. The failure was formatting.

I think the last thing to do about this PR is to format the code (according to https://github.com/pingcap/tiflash#contributing) and let the CI pass. Then it will lgtm.

@Jiang-Hua

Jiang-Hua commented:
> I checked the latest CI build. The cpp build seems fine. The failure was formatting.
> 
> I think the last thing to do about this PR is to format the code (according to https://github.com/pingcap/tiflash#contributing) and let the CI pass. Then it will lgtm.
> 
> @Jiang-Hua

Got it and will fix issues, thanks.

Lloyd-Pottiger commented:
/run-all-tests

Lloyd-Pottiger commented:
@Jiang-Hua should we check out to a stable version? like [v1.1.0](https://github.com/intel/qpl/releases/tag/v1.1.0)

zanmato1984 reviewed (approved):
LGTM

zanmato1984 reviewed (changes requested):
One remaining concern. Withdrawing my lgtm.

zanmato1984 reviewed (changes requested):
One remaining concern. Withdrawing my lgtm.

zanmato1984 reviewed (approved):
LGTM

zanmato1984 reviewed (approved):
LGTM

zanmato1984 commented:
/run-all-tests

Lloyd-Pottiger commented:
@Jiang-Hua Thanks for your excellent work! Could you please sign cla https://cla.pingcap.net/pingcap/tiflash?pullRequest=7364 so that we are able to merge it?

Lloyd-Pottiger commented:
/run-integration-test

zanmato1984 commented:
/run-all-tests

zanmato1984 commented:
/approve

Lloyd-Pottiger commented:
/unhold
